tests/server.py

```diff
@@ -602,30 +602,48 @@ def _make_test_homeserver_synchronous(server: HomeServer) -> None:
     for database in server.get_datastores().databases:
         pool = database._db_pool
 
-        def runWithConnection(func, *args, **kwargs):
-            return threads.deferToThreadPool(
+        async def runWithConnection(
+            func: Callable[..., R],
```

Comment:
I think turning this into a coroutine might have broken something, judging by:

```
dmr on titan in synapse on  clokep/server-types is 📦 v1.77.0 via 🐍 v3.11.1 (matrix-synapse-py3.11) via 🦀 v1.63.0 took 3s 
2023-02-16 20:52:41 ✔  $ trial tests.api.test_auth.AuthTestCase.test_blocking_mau | tail -n 20



tests.api.test_auth.AuthTestCase.test_blocking_mau
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/dmr/.cache/pypoetry/virtualenvs/matrix-synapse-YNqvS2_4-py3.11/lib/python3.11/site-packages/twisted/trial/_asynctest.py", line 205, in deferRunCleanups
    yield func(*args, **kwargs)
  File "/home/dmr/workspace/synapse/tests/test_utils/__init__.py", line 96, in cleanup
    raise exc
  File "/usr/lib64/python3.11/warnings.py", line 537, in _warn_unawaited_coroutine
    warn(msg, category=RuntimeWarning, stacklevel=2, source=coro)
builtins.RuntimeWarning: coroutine '_make_test_homeserver_synchronous.<locals>.runWithConnection' was never awaited

tests.api.test_auth.AuthTestCase.test_blocking_mau
-------------------------------------------------------------------------------
Ran 1 tests in 0.151s

FAILED (errors=2)
```

Reply:
Ohhhhh I bet because `Deferreds` run automatically while awaitables don't or whatever?

Reply:
I ended up simplifying these slightly so they don't match the exact signature.